modules/tinymce/src/core/main/ts/api/Options.ts

```diff
@@ -0,0 +1,810 @@
+/**
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Arr, Obj, Strings, Type } from '@ephox/katamari';
+import { PlatformDetection } from '@ephox/sand';
+
+import DOMUtils from './dom/DOMUtils';
+import Editor from './Editor';
+import { EditorOptions } from './OptionTypes';
+import I18n from './util/I18n';
+import Tools from './util/Tools';
+
+const deviceDetection = PlatformDetection.detect().deviceType;
+const isTouch = deviceDetection.isTouch();
+const DOM = DOMUtils.DOM;
+
+const getHash = (value: string): Record<string, string> => {
+  const items = value.indexOf('=') > 0 ? value.split(/[;,](?![^=;,]*(?:[;,]|$))/) : value.split(',');
+  return Arr.foldl(items, (output, item) => {
+    const arr = item.split('=');
+    const key = arr[0];
+    const val = arr.length > 1 ? arr[1] : key;
+    output[Strings.trim(key)] = Strings.trim(val);
+    return output;
+  }, {} as Record<string, string>);
+};
+
+const option = <K extends keyof EditorOptions>(name: K) => (editor: Editor) =>
+  editor.options.get(name);
+
+const stringOrObjectProcessor = (value: string) =>
+  Type.isString(value) || Type.isObject(value);
+
+const bodyOptionProcessor = (editor: Editor, defaultValue: string = '') => (value: string) => {
+  const valid = Type.isString(value);
+  if (valid) {
+    if (value.indexOf('=') !== -1) {
+      const bodyObj = getHash(value);
+      return { value: Obj.get(bodyObj, editor.id).getOr(defaultValue), valid };
+    } else {
+      return { value, valid };
+    }
+  } else {
+    return { value, valid, message: 'Must be a string.' };
+  }
+};
+
+const register = (editor: Editor) => {
+  const registerOption = editor.options.register;
+
+  registerOption('id', {
```

Comment:
I wonder if we should have a bulk register function instead or if register should be able to take just a object. So you could do.
```
editor.options.register({
  selector: { processor: 'string' },
  target: { processor: 'object' },
  referrer_policy: { processor: 'string', default: '' }
})
```
It would be far less verbose if you have many settings like this. If we don't want it in the public api then we should at least consider it for this file since it will be come pretty large. It could also be reduced by having predefined register functions for the common types. For this large file might be worth having things like `registerString('id', editor.id)` to avoid having a object structure for every single option.

Reply:
We had discussed a `registerAll` function like this but opted to leave it out until actually needed. Happy to re-evaluate and include that now if needed (though the types for that could be nasty).

As for `registerString` we hadn't mentioned this specific case, but in general we specifically don't want to be using function arguments like that as it severely locks in what properties you can use. An object we can continuously add mandatory or optional properties, but that can't be done when its function arguments without making a breaking change.

Reply:
Yes both a registerAll or a registerString can be local functions to the editor options module. The registerString was more of an alternative to the full register with all properties. Most options will be simple things boolean, strings etc so it seems very verbose to have a config for these simple things as you can editor default options file shows it becomes a lot of duplicated similar objects. Like { processor: 'boolean' } 30 times. But I guess most can be worked around by local functions like the option function I see here and there in this PR.

Reply:
I wonder if instead we could have a public API such as `registerString(name: string, spec: Omit<BuiltInOptionSpec, 'processor'>)` (or similar). That way you wouldn't have to define the processor every time, but would still alleviate some of the syntax duplication...

Anyways, you're right we can use an internal function for big files such as this. We may end up with some duplication elsewhere though amongst the individual plugins though so I'll discuss the above more before going down that path.

Reply:
Discussed this at stand up and the team we're keen on adding public APIs just yet. We also don't really gain anything from the internal `registerAll` function and instead are potentially going to lose type information. So for now we're going to keep this as is and re-evaluate once we've converted more before we do the final merge into `develop`.